horovod/common/controller.h

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef HOROVOD_CONTROL_MANAGER_H
+#define HOROVOD_CONTROL_MANAGER_H
+
+#include <iostream>
+#include <queue>
+#include <vector>
+
+#include "half.h"
+#include "message_table.h"
+#include "parameter_manager.h"
+#include "response_cache.h"
+#include "stall_inspector.h"
+#include "tensor_queue.h"
+
+#if HAVE_GLOO
+#include "gloo_context.h"
+#endif
+
+namespace horovod {
+namespace common {
+
+// Forward declaration
+class HorovodGlobalState;
+
+class Controller : public std::enable_shared_from_this<Controller> {
+public:
+  Controller(HorovodGlobalState& global_state);
```

Comment:
Can we actually pass args explicitly and not make Controller aware of HorovodGlobalState existence?

Reply:
Yes.